### PR TITLE
Handling Facebook errors

### PR DESCRIPTION
**Pull request by depado**
…ion calls

Also quick typo in the comments of the `Read` struct. 
I didn't include much detail in the error message (the code is already in the message anyway), and I have no idea what the `FbtraceID` is about ^^

Cheers

### CODE REVIEW

File: response.go [hunk 1]
```diff
@@ -16,6 +16,21 @@ const (
...
+	Error  *QueryError `json:"error,omitempty"`
+	Result string      `json:"result,omitempty"`
+}
+
+// QueryError is representing an error sent back by facebook
+type QueryError struct {
+	Message   string `json:"message"`
+	Type      string `json:"type"`
+	Code      int    `json:"code"`
+	FbtraceID string `json:"fbtrace_id"`
```
Comment (paked) [inline response.go]: Let's rename this field to `FBTraceID`

File: response.go [hunk 2]
```diff
@@ -16,6 +16,21 @@ const (
 	SendMessageURL = "https://graph.facebook.com/v2.6/me/messages"
 )
 
+// QueryResponse is the response sent back by facebook when setting up things
```
Comment (paked) [inline response.go]: Capitalise Facebook

File: response.go [hunk 3]
```diff
@@ -16,6 +16,21 @@ const (
...
+// like greetings or call-to-actions
+type QueryResponse struct {
+	Error  *QueryError `json:"error,omitempty"`
+	Result string      `json:"result,omitempty"`
+}
+
+// QueryError is representing an error sent back by facebook
```
Comment (paked) [inline response.go]: Capitalise Facebook

**Review by paked — CHANGES_REQUESTED** [review]
Sorry it took me a while to get back to this.

This is heading in the right direction, I've left a few comments regarding minor stuff like documentation etc.

After you've fixed those errors, It'd be great if you could take a stab at handling these errors in other places. I imagine that they show up in other places too, like:
- https://github.com/paked/messenger/blob/master/messenger.go#L125
- https://github.com/paked/messenger/blob/master/response.go#L139
- https://github.com/paked/messenger/blob/master/response.go#L175
- https://github.com/paked/messenger/blob/master/response.go#L203

... And probably more.

Let me know if you're up to it, if not we can just work on getting these merged.

**Comment by paked** [comment]
Also, on another note: The commit you've sent doesn't seem to have an author associated with it. [I'd follow the instructions here](https://help.github.com/articles/set-up-git/#setting-up-git). Once you've done that, `git commit --amend && git push -f` should fix everything up!

**Comment by depado** [comment]
Hello there ! Thanks for the feedback, I'll have a look at this today.

**Comment by depado** [comment]
Hey there. 
So I made the changes you requested (comments and field name) and also added this behavior for the other functions that can potentially fail. 

Also I fixed my author thing I think. 

Does it look good to you ?

**Review by paked — CHANGES_REQUESTED** [review]
This looks great. Before I merge it in, I'd like to reduce the amount of code repetition from this new Facebook error handling code.

Do you think you could move it into something like`func checkFacebookError(r io.Reader) err`?

After this is done we'll be able to merge this straight in.

**Comment by depado** [comment]
Sure thing :)

File: messenger.go [hunk 4]
```diff
@@ -122,7 +122,11 @@ func (m *Messenger) ProfileByID(id int64) (Profile, error) {
 	resp, err := client.Do(req)
 	defer resp.Body.Close()
 
-	err = json.NewDecoder(resp.Body).Decode(&p)
```
Comment (depado) [inline messenger.go:125]: That would break everything :smile:

**Comment by depado** [comment]
There's an issue with the `ProfileByID` function as the returned JSON is either the profile of the user or an error. (So it can either be a `Profile` or a `QueryResponse` type) I'll give that a try later. Thanks for the feedback :)

File: messenger.go (hunk 4 above)
Comment (paked) [reply messenger.go:125]: 😆

**Comment by paked** [comment]
Alright. Good luck!

**Comment by depado** [comment]
Hey there. Found a way to handle that kind of corner case. Could you review the changes ? 
It works pretty fine from what I can see (for example try to `ProfileByID(0)` it will return an error).

**Comment by paked** [comment]
@Depado Thanks for your persistence on this PR! I'm very grateful 🎈

**Pull request merged**